planning/freespace_planning_algorithms/src/astar_search.cpp

```diff
@@ -137,17 +139,8 @@ void AstarSearch::setMap(const nav_msgs::msg::OccupancyGrid & costmap)
 
   clearNodes();
 
-  const auto height = costmap_.info.height;
-  const auto width = costmap_.info.width;
-
-  // Initialize nodes
-  nodes_.resize(height);
-  for (size_t i = 0; i < height; i++) {
-    nodes_[i].resize(width);
-    for (size_t j = 0; j < width; j++) {
-      nodes_[i][j].resize(planner_common_param_.theta_size);
-    }
-  }
+  x_scale_ = costmap_.info.height;
+  graph_.reserve(100000);
```

Comment:
How did you choose the value `100000` ? Can it be calculated from the `costmap` dimensions and the `y_scale_` ?

Reply:
I did considered to set the value dynamically according to map parameters but the nodes actually needed varies largely in different cases even using the same map. I've tested different cases in this [discussion](https://github.com/orgs/autowarefoundation/discussions/2716#discussion-4207199). The "Visited-nodes" column corresponds to nodes quantity. The number could be larger in low-resolution map than in high-resolution map if the case using low-resolution map is complicated enough.

On the other hand, the initial scale of graph_ seems not really allocated in memory. So maybe it will not have an obvious influence on efficiency.

By the way, could you please review this PR?

Reply:
I am sorry the discussion you opened went unnoticed. The observations you made are very good and your PR makes a lot of sense. Once this is merged, you can post a link of the PR in the discussion and mark it as the answer.

In the future, if people do not react to your discussion, you can notify people during the software working group or on Discord.

Reply:
> I understand now, thank you. I think in that case it is better to not `reserve` here.

I compared settings with and without "graph_.reserve(n)". I think it's better to keep the reserve action cause it can make the performance more stable. Test results are as follow.

A complicated case: nearly 300,000 nodes are expanded.
- start point: "pose: {position: {x: 3713.99, y: 73732.0, z: 0.664812}, orientation: {x: 0.0, y: 0.0, z: 0.850511, w: 0.525957}"
- goal point: "pose: {position: {x: 3746.88, y: 73748.5, z: 0.0}, orientation: {x: 0.0, y: 0.0, z: 0.847164, w: 0.531332}"

![image](https://user-images.githubusercontent.com/103234047/195003523-941b5b14-5264-498f-876d-82fb0d61fd21.png)

![image](https://user-images.githubusercontent.com/103234047/195005869-a967f7f6-b834-4f98-aef1-674e8af8b301.png)

From the table above, we can conclude:
- Without pre-allocation, it may cost more time due to reasons like frequent rehash.
- Smaller pre-allocation size has little influence on general performance. (There may exist a key size that is sensitive to performance, still need testing.)

Reply:
> I am sorry the discussion you opened went unnoticed. The observations you made are very good and your PR makes a lot of sense. Once this is merged, you can post a link of the PR in the discussion and mark it as the answer.
> 
> In the future, if people do not react to your discussion, you can notify people during the software working group or on Discord.

I don't know why the discussion link is not permanent. Thanks for your suggestion. I will paste this PR to the discussion later.